File: aulas/aula_07/q4-cinema.py
"""
Aula 07 - Encapsulamento
"""

class Ingresso:
    __dia = ""
    __horario = None
    def __init__(self):
        self.__dia = str
        self.__horario = int

    # Encapsulamento
    def get_dia(self):
        return self.__dia
    
    def get_horario(self):
        return self.__horario
    
    def set_dia(self, dia):
        if (dia in ["seg", "ter", "qua", "qui", "sex", "sab", "dom"]):
            self.__dia = dia
        else:
            raise ValueError("Dia inválido")

    def set_horario(self, horario):
        if (horario >= 0 and horario <= 23):
            self.__horario = horario
        else:
            raise ValueError("Hora Inválida")
    # Fim encapsulamento

    def inteira(self):
        dia = self.__dia
        horario = self.__horario
        if dia in ["seg", "ter", "qui"]:
            valor = 16
            if horario == 0 or horario >= 17:
                return valor * 1.5
            return valor
        
        elif dia == "qua":
            return 8
        
        elif dia in ["sex", "sab", "dom"]:
            valor = 20
            if horario == 0 or horario >= 17:
                return valor * 1.5
            return valor
    
    def meia(self):
        dia = self.__dia
        if dia == "qua":
            return 8
        return self.inteira()/2


class UI:
    def main():
        op = 0
        while op != 2:
            op = UI.menu()
            if op == 1:
                UI.ingresso()
        print("Saindo...")
        
    def menu():
        opcoes = ["1 - Ingressos", "2 - Sair"]

        for i in opcoes:
            print(i)

        op = int(input("Escolha uma opção: "))
        return op
    
    def ingresso():
        dia = input("Informe o dia: ")
        horario = int(input("Informe o horário: "))
        ingr = Ingresso()
        ingr.set_dia(dia)
        ingr.set_horario(horario) 
        preco_inteira = ingr.inteira()
        preco_meia = ingr.meia()
        print(f"Preço da inteira: R${preco_inteira:0.2f}")
        print(f"Preço da meia: R${preco_meia:0.2f}")

UI.main()

    

File: listas/lista01/q4-cinema.cpp
#include <iostream>
using namespace std;

class Ingresso {
    public:
    string dia;
    short hora;

    Ingresso(string dia_arg, short hora_arg) {
        dia = dia_arg;
        hora = hora_arg;
    }

    int calcula_ingresso() {
        if (dia == "seg" || dia == "ter" || dia == "qui") {
            int valor = 16;
            if (hora == 0 || hora >= 17) {
                valor = valor * 1.5;
            }

            return valor;
        } 
        else if (dia == "qua") {
            int valor = 8;
            return valor;
        }
        else if (dia == "sex" || dia == "sab" || dia == "dom") {
            int valor = 20;
            if (hora == 0 || hora >= 17) {
                valor = valor * 1.5;
            }
            return valor;
        }
    }
};

class UI {
    public:
    string cabecalho;

    UI() {
        cabecalho = "========= Bem-vindo(a) ao seletor de ingressos =========";
    }

    void main() {
        cout << cabecalho << endl;
        int op = 0;
        while (op != 2) {
            op = this->menu();
            if (op == 1) {
                this->ingresso();
            };
        };
        cout << "========= Até a próxima =========" << endl;
    }

    int menu() {
        int op;
        short ARRAY_LENGTH = 2;
        string opcoes[ARRAY_LENGTH] = {"1 - Ingressos", "2 - Sair"};

        for (int i=0; i < ARRAY_LENGTH; ++i) {
            cout << opcoes[i] << endl;
        };

        cout << "Escolha um valor:" << endl;
        cin >> op;
        return op;
    }

    void ingresso() {
        string dia;
        short hora;
        int inteira;
        cout << "Informe o dia: ";
        cin >> dia;
        cout << "Informe a hora: ";
        cin >> hora;
        Ingresso ingr(dia, hora);
        
        inteira = ingr.inteira();
        cout << "Preço da inteira: R$" << inteira << endl;
    }
};

int main() {
    UI ui;
    ui.main();
    return 0;
}